### PR TITLE
[GenAPI] Synthesize an internal default constructor for the type with implicit default constructor and the base type without.

### LINKED ISSUE #12578: Synthesize an internal default constructor for the type with implicit default constructor and the base type without.
```C#
public class Bar
{
    public Bar(int a) { }
}

public class Foo : Bar
{
/// Synthesize an internal default constructor
    internal Foo() : base(1) { }
}
```

### PR DESCRIPTION
**Pull request by andriipatsula**
Fixes: https://github.com/dotnet/arcade/issues/12578

### CODE REVIEW

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs [hunk 1]
```diff
@@ -159,5 +160,54 @@ public static bool IsBoundGenericType(this INamedTypeSymbol namedType)
...
+        {
+            if (!namedType.Constructors.Where(symbolFilter.Include).Any() && namedType.BaseType != null)
+            {
+                IEnumerable<IMethodSymbol> baseConstructors = namedType.BaseType.Constructors.Where(symbolFilter.Include);
+                if (baseConstructors.Any() && baseConstructors.All(c => !c.Parameters.IsEmpty))
+                {
+                    SyntaxKind visibility = SyntaxKind.InternalKeyword;
+
+                    if ((symbolFilter is AccessibilitySymbolFilter accessibilityFilter && accessibilityFilter.IncludeInternalSymbols) ||
+                        (symbolFilter is CompositeSymbolFilter compositeSymbolFilter &&
```
Comment (smasher164) [inline src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:175]: Maybe it'd be good to extract this check into a function, like `IncludeInternalSymbols(symbolFilter)`?

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs [hunk 2]
```diff
@@ -159,5 +160,54 @@ public static bool IsBoundGenericType(this INamedTypeSymbol namedType)
...
+                        visibility = SyntaxKind.PrivateKeyword;
+                    }
+
+                    yield return SyntaxFactory.ConstructorDeclaration(
+                        new SyntaxList<AttributeListSyntax>(),
+                        SyntaxFactory.TokenList(new[] { SyntaxFactory.Token(visibility) }),
+                        SyntaxFactory.Identifier(namedType.ToDisplayString()),
+                        SyntaxFactory.ParameterList(),
+                        default!,
+                        default(BlockSyntax)!).WithInitializer(baseConstructors.First().GenerateBaseConstructorInitializer()); ;
```
Comment (smasher164) [inline src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:188]: extra semicolon here

**Review by smasher164 — APPROVED** [review]

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs (hunk 2 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs]: Updated.

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs (hunk 1 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:178]: Introduced a lambda function:

```C#
Func<ISymbolFilter, bool> IncludeInternalSymbols = filter =>
    filter is AccessibilitySymbolFilter accessibilityFilter && accessibilityFilter.IncludeInternalSymbols;
```

File: src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/AccessibilitySymbolFilter.cs [hunk 3]
```diff
@@ -14,6 +14,9 @@ public class AccessibilitySymbolFilter : ISymbolFilter
         private readonly bool _includeEffectivelyPrivateSymbols;
         private readonly bool _includeExplicitInterfaceImplementationSymbols;
 
+        public bool IncludeInternalSymbols { get => _includeInternalSymbols; }
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/AccessibilitySymbolFilter.cs:17]:
```suggestion
        public bool IncludeInternalSymbols { get; }
```

and remove the underlying backing field as the compiler generates one automatically. Same for `IncludeEffectivelyPrivateSymbols`

File: src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/CompositeSymbolFilter.cs [hunk 4]
```diff
@@ -15,6 +15,8 @@ public class CompositeSymbolFilter : ISymbolFilter
     {
         private readonly List<ISymbolFilter> _innerFilters = new();
 
+        public List<ISymbolFilter> Filters { get => _innerFilters; }
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/CompositeSymbolFilter.cs:18]: Same as in AccessibilitySymbolFilter, remove the backing field and let the compiler generate it automatically:

```suggestion
        public List<ISymbolFilter> Filters { get; }
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs [hunk 5]
```diff
@@ -159,5 +160,56 @@ public static bool IsBoundGenericType(this INamedTypeSymbol namedType)
 
             return true;
         }
+
+        // Synthesize an internal default constructor for the type with implicit default constructor and the base type without.
+        public static IEnumerable<SyntaxNode> TryGetInternalDefaultConstructor(this INamedTypeSymbol namedType, ISymbolFilter symbolFilter)
+        {
+            if (!namedType.Constructors.Where(symbolFilter.Include).Any() && namedType.BaseType != null)
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:167]: Switch the condition checks as the `namedType.BaseType != null` is O(1) while the former condition is at least O(n).

Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:167]:
```suggestion
            if (!namedType.Constructors.Any(symbolFilter.Include) && namedType.BaseType != null)
```

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs [hunk 6]
```diff
@@ -1389,5 +1389,163 @@ public partial class foo
...
+                    namespace A
+                    {
+                        public partial class Bar
+                        {
+                            public Bar(int a) { }
+                        }
+
+                        public partial class Foo : Bar
+                        {
+                            internal Foo() : base(default) { }
```
Comment (ViktorHofer) [inline src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1432]: Maybe a naive question but why don't we provide 1 as the parameter and instead use `default`?

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs [hunk 7]
```diff
@@ -1389,5 +1389,163 @@ public partial class foo
                     """,
                 includeInternalSymbols: false);
         }
+
+        [Fact]
+        public void TestInternalDefaultConstructorGeneration()
```
Comment (ViktorHofer) [inline src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1394]: Please add a test with `includeInternalSymbols = true`. I would expect that such a test would show that a default ctor would become a private one.

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs [hunk 8]
```diff
@@ -159,5 +160,56 @@ public static bool IsBoundGenericType(this INamedTypeSymbol namedType)
...
+            {
+                IEnumerable<IMethodSymbol> baseConstructors = namedType.BaseType.Constructors.Where(symbolFilter.Include);
+                if (baseConstructors.Any() && baseConstructors.All(c => !c.Parameters.IsEmpty))
+                {
+                    SyntaxKind visibility = SyntaxKind.InternalKeyword;
+
+                    Func<ISymbolFilter, bool> IncludeInternalSymbols = filter =>
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:174]: nit: lambda function variables are pascal-cased: `includeInternalSymbols`.

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs [hunk 9]
```diff
@@ -159,5 +160,56 @@ public static bool IsBoundGenericType(this INamedTypeSymbol namedType)
...
+                        SyntaxFactory.TokenList(new[] { SyntaxFactory.Token(visibility) }),
+                        SyntaxFactory.Identifier(namedType.ToDisplayString()),
+                        SyntaxFactory.ParameterList(),
+                        default!,
+                        default(BlockSyntax)!).WithInitializer(baseConstructors.First().GenerateBaseConstructorInitializer());
+                }
+            }
+        }
+
+        // Synthesize a base class initializer.
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:195]: As this API is publicly exposed, please use triple-slash documentation here: `///`

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs [hunk 10]
```diff
@@ -159,5 +160,56 @@ public static bool IsBoundGenericType(this INamedTypeSymbol namedType)
 
             return true;
         }
+
+        // Synthesize an internal default constructor for the type with implicit default constructor and the base type without.
```
Comment (ViktorHofer) [inline src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs:164]: As this API is publicly exposed, please use triple-slash documentation here: ///

Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs]: Updated

```C#
/// <summary>
/// Synthesize an internal default constructor for the type with implicit default constructor and the base type without.
/// </summary>
/// <param name="namedType">A loaded named type symbol <see cref="INamedTypeSymbol"/>.</param>
/// <param name="symbolFilter">Assembly symbol filter <see cref="ISymbolFilter"/>.</param>
```

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs (hunk 8 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs]: Thanks, updated.

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs [hunk 11]
```diff
@@ -159,5 +160,56 @@ public static bool IsBoundGenericType(this INamedTypeSymbol namedType)
...
+            {
+                IEnumerable<IMethodSymbol> baseConstructors = namedType.BaseType.Constructors.Where(symbolFilter.Include);
+                if (baseConstructors.Any() && baseConstructors.All(c => !c.Parameters.IsEmpty))
+                {
+                    SyntaxKind visibility = SyntaxKind.InternalKeyword;
+
+                    Func<ISymbolFilter, bool> IncludeInternalSymbols = filter =>
+                        filter is AccessibilitySymbolFilter accessibilityFilter && accessibilityFilter.IncludeInternalSymbols;
+
+                    if (IncludeInternalSymbols(symbolFilter) ||
```
[the remark this answers is not in the document]
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs]: Added:
> // We use the `Private` visibility if internal symbols are not filtered out.

File: src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs (hunk 9 above)
Comment (andriipatsula) [reply src/GenAPI/Microsoft.DotNet.GenAPI/INamedTypeSymbolExtension.cs]: Updated.
```
/// <summary>
/// Synthesize a base class initializer. 
/// </summary>
/// <param name="baseTypeConstructor">Represents a base class constructor <see cref="IMethodSymbol"/>.</param>
/// <returns></returns>
```

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs (hunk 6 above)
Comment (andriipatsula) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1432]: I don't think we have access to parameter of the base class constructor call as this is part of implementation details. `internal Foo() : base(1) { }`. Technically, this doesn't meter and `default` is universal solution.

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs (hunk 7 above)
Comment (andriipatsula) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1394]: `includeInternalSymbols` is true by default. Explicitly specify it in the `TestPrivateDefaultConstructorGeneration` test case

File: src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/AccessibilitySymbolFilter.cs [hunk 12]
```diff
@@ -10,22 +10,22 @@ namespace Microsoft.DotNet.ApiSymbolExtensions.Filtering
     /// </summary>
     public class AccessibilitySymbolFilter : ISymbolFilter
     {
-        private readonly bool _includeInternalSymbols;
-        private readonly bool _includeEffectivelyPrivateSymbols;
-        private readonly bool _includeExplicitInterfaceImplementationSymbols;
+        public bool IncludeExplicitInterfaceImplementationSymbols { get; }
+        public bool IncludeInternalSymbols { get; }
+        public bool IncludeEffectivelyPrivateSymbols { get; }
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/AccessibilitySymbolFilter.cs:15]: As this is public API, please add triple slash XML documentation for each property.

File: src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs (hunk 7 above)
Comment (ViktorHofer) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1394]: Do we have a test that shows that the implicit default ctor becomes private?

Comment (andriipatsula) [reply src/Tests/Microsoft.DotNet.GenAPI.Tests/CSharpFileBuilderTests.cs:1394]: > implicit default ctor becomes private?

Could you please provide a short example?

In the example below:

```C#
public class Bar
{
/// implicit default constructor.
}
```

we won't generate `private Bar() {}` even if the `includeInternalSymbols` is set to true.

File: src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/AccessibilitySymbolFilter.cs (hunk 12 above)
Comment (andriipatsula) [reply src/Microsoft.DotNet.ApiSymbolExtensions/Filtering/AccessibilitySymbolFilter.cs]: updated

**Pull request merged**